### PR TITLE
better BPINN formulation, improvements to include dataset domain points

### PR DESCRIPTION
**Pull request by AstitvaAggarwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context

will add docs and more tests separately

### CODE REVIEW

**Comment by AstitvaAggarwal** [comment]
@Vaibhavdixit02 @ChrisRackauckas this is the new pr, closed the previous one due to large number of merge conflicts after the repo was recently majorly formatted.

File: src/collocated_estim.jl [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        #         Tar.prob.p,
+        #         t[i])
+        #              for i in 1:length(û[:, 1])]
+        # end
+        # form of NN output matrix output dim x n
+        # nnsol = reduce(hcat, nnsol)
+
+        # > Instead of dataset gradients trying NN derivatives with dataset collocation 
+        # # convert to matrix as nnsol  
+
```
Comment (ChrisRackauckas) [inline src/collocated_estim.jl:91]: these long comments should get cleaned up

File: src/collocated_estim.jl [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+#         derivative = (x̂[i + 1] - x̂[i - 1]) / (Δt_forward + Δt_backward)
+
+#         derivatives[i] = derivative
+#     end
+
+#     # Derivatives at the endpoints can be calculated using forward or backward differences.
+#     derivatives[1] = (x̂[2] - x̂[1]) / (time[2] - time[1])
+#     derivatives[end] = (x̂[end] - x̂[end - 1]) / (time[end] - time[end - 1])
+#     return derivatives
+# end
```
Comment (ChrisRackauckas) [inline src/collocated_estim.jl:130]: What's this all about?

File: src/collocated_estim.jl [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    #   chainflux(dataset[end]' .+ sqrt(eps(eltype(Float64))))')
+
+    # A2 = (prob.u0' .+
+    #       (prob.tspan[2] .- (dataset[end]'))' .*
+    #       chainflux(dataset[end]')')
+
+    A1 = chainflux(dataset[end]' .+ sqrt(eps(eltype(dataset[end][1]))))
+    A2 = chainflux(dataset[end]')
+
+    gradients = (A2 .- A1) ./ sqrt(eps(eltype(dataset[end][1])))
```
Comment (ChrisRackauckas) [inline src/collocated_estim.jl:162]: You shouldn't use this for gradients?

**Comment by ChrisRackauckas** [comment]
There's a lot of random stuff together in here. Break this apart into simple one thing PRs and lets merge those.

File: src/collocated_estim.jl (hunk 3 above)
Comment (AstitvaAggarwal) [reply src/collocated_estim.jl:162]: no, we use the gradient of the NN in new loss, ill remove this.

Comment (AstitvaAggarwal) [reply src/collocated_estim.jl:162]: not anymore, i wrote this for it to work for a specific case, normally the same is done using the derivative function NNodederi in the BPINN ode solver.

File: src/collocated_estim.jl (hunk 2 above)
Comment (AstitvaAggarwal) [reply src/collocated_estim.jl:130]: initially wrote this for testing, while figuring out which kind of derivative is to be used in new loss. the above was the case where the derivative was datapoint interpolation derivatives via central diff.

File: src/collocated_estim.jl (hunk 1 above)
Comment (AstitvaAggarwal) [reply src/collocated_estim.jl:91]: okay

**Comment by Vaibhavdixit02** [comment]
@AstitvaAggarwal do you think you can make this into multiple smaller PRs as Chri suggested above?

**Comment by AstitvaAggarwal** [comment]
Yes, will do

**Comment by AstitvaAggarwal** [comment]
@ChrisRackauckas @Vaibhavdixit02 I've seperated the code into two PRs. one for the better ode solver and this for the PDE solver.

**Comment by ChrisRackauckas** [comment]
this one is still having issues with tests passing?

**Comment by AstitvaAggarwal** [comment]
tests should pass. also turns out improvement is seen at any added noise level.

File: Project.toml [hunk 4]
```diff
@@ -75,7 +75,7 @@ Reexport = "1.2"
 RuntimeGeneratedFunctions = "0.5.12"
 SafeTestsets = "0.1"
 SciMLBase = "2.28"
-Statistics = "1.10"
```
Comment (avik-pal) [inline Project.toml:78]: accidental?

File: src/PDE_BPINN.jl [hunk 5]
```diff
@@ -58,15 +63,85 @@ mutable struct PDELogTargetDensity{
...
+
+    return to_subs, tobe_subs
+end
+
 function LogDensityProblems.logdensity(Tar::PDELogTargetDensity, θ)
     # for parameter estimation neccesarry to use multioutput case
-    return Tar.full_loglikelihood(setparameters(Tar, θ),
-               Tar.allstd) + priorlogpdf(Tar, θ) + L2LossData(Tar, θ)
-    # + L2loss2(Tar, θ)
+    if Tar.L2_loss2 isa Nothing
```
Comment (avik-pal) [inline src/PDE_BPINN.jl:137]:
```suggestion
    if Tar.L2_loss2 === nothing
```

**Review by avik-pal — COMMENTED** [review]
do a rebase

**Comment by avik-pal** [comment]
Maybe copy over your code into a new PR, this seems to have messed up the history (the diff from old PRs are showing up here)

**Comment by AstitvaAggarwal** [comment]
Yeah lol, ive been on it today and yesterday

**Pull request closed without merging**